crypto/crmf/crmf_lib.c

```diff
@@ -0,0 +1,753 @@
+IMPLEMENT_CRMF_CTRL_FUNC(protocolEncrKey, X509_PUBKEY, regCtrl)
+
+/*
+ * Pushes the attribute given in regInfo in to the CertReqMsg->regInfo stack.
+ * (section 7)
+ * returns 1 on success, 0 on error
+ */
+static int OSSL_CRMF_MSG_push0_regInfo(OSSL_CRMF_MSG *crm,
+                                       OSSL_CRMF_ATTRIBUTETYPEANDVALUE *ri)
+{
+    int new = 0;
+
+    if (crm == NULL || ri == NULL) {
+        CRMFerr(CRMF_F_OSSL_CRMF_MSG_PUSH0_REGINFO, CRMF_R_NULL_ARGUMENT);
+        return 0;
+    }
+
+    if (crm->regInfo == NULL) {
+        if ((crm->regInfo =
+             sk_OSSL_CRMF_ATTRIBUTETYPEANDVALUE_new_null()) == NULL)
+            goto oom;
+        new = 1;
+    }
+    if (!sk_OSSL_CRMF_ATTRIBUTETYPEANDVALUE_push(crm->regInfo, ri))
+        goto oom;
+    return 1;
+ oom:
+    CRMFerr(CRMF_F_OSSL_CRMF_MSG_PUSH0_REGINFO, ERR_R_MALLOC_FAILURE);
+
+    if (new != 0) {
+        sk_OSSL_CRMF_ATTRIBUTETYPEANDVALUE_free(crm->regInfo);
+        crm->regInfo = NULL;
+    }
+    return 0;
+}
+
+ /* id-regInfo-utf8Pairs to regInfo (section 7.1) */
+IMPLEMENT_CRMF_CTRL_FUNC(utf8Pairs, ASN1_UTF8STRING, regInfo)
+
+ /* id-regInfo-certReq to regInfo (section 7.2) */
+IMPLEMENT_CRMF_CTRL_FUNC(certReq, OSSL_CRMF_CERTREQUEST, regInfo)
+
+
+/* retrieves the certificate template of crm */
+OSSL_CRMF_CERTTEMPLATE *OSSL_CRMF_MSG_get0_tmpl(const OSSL_CRMF_MSG *crm)
+{
+    if (crm == NULL || crm->certReq == NULL) {
+        CRMFerr(CRMF_F_OSSL_CRMF_MSG_GET0_TMPL, CRMF_R_NULL_ARGUMENT);
+        return NULL;
+    }
+    return crm->certReq->certTemplate;
+}
+
+
+int OSSL_CRMF_MSG_set_validity(OSSL_CRMF_MSG *crm, time_t from, time_t to)
+{
+    OSSL_CRMF_OPTIONALVALIDITY *vld = NULL;
+    ASN1_TIME *from_asn = NULL;
+    ASN1_TIME *to_asn = NULL;
+    OSSL_CRMF_CERTTEMPLATE *tmpl = OSSL_CRMF_MSG_get0_tmpl(crm);
+
+    if (tmpl == NULL) { /* also crm == NULL implies this */
+        CRMFerr(CRMF_F_OSSL_CRMF_MSG_SET_VALIDITY, CRMF_R_NULL_ARGUMENT);
+        return 0;
+    }
+
+    if (from != 0 && ((from_asn = ASN1_TIME_set(NULL, from)) == NULL))
+        goto oom;
+    if (to != 0 && ((to_asn = ASN1_TIME_set(NULL, to)) == NULL))
+        goto oom;
+    if ((vld = OSSL_CRMF_OPTIONALVALIDITY_new()) == NULL)
+        goto oom;
+
+    vld->notBefore = from_asn;
+    vld->notAfter = to_asn;
+
+    tmpl->validity = vld;
+
+    return 1;
+ oom:
+    CRMFerr(CRMF_F_OSSL_CRMF_MSG_SET_VALIDITY, ERR_R_MALLOC_FAILURE);
+    ASN1_TIME_free(from_asn);
+    ASN1_TIME_free(to_asn);
+    return 0;
+}
+
+
+int OSSL_CRMF_MSG_set_certReqId(OSSL_CRMF_MSG *crm, int rid)
+{
+    if (crm == NULL || crm->certReq == NULL || crm->certReq->certReqId == NULL) {
+        CRMFerr(CRMF_F_OSSL_CRMF_MSG_SET_CERTREQID, CRMF_R_NULL_ARGUMENT);
+        return 0;
+    }
+
+    return ASN1_INTEGER_set(crm->certReq->certReqId, rid);
+}
+
+/* get ASN.1 encoded integer, return -1 on error */
+static int CRMF_ASN1_get_int(int func, const ASN1_INTEGER *a)
```

Comment:
for a local function the name should be lowercase, like crmf_asn1_get_int

Reply:
changed